### PR TITLE
Credential API

### PR DESCRIPTION
**Pull request by ushergod**
GET /credentials: get credentials by credentials file path
PUT /credentials: update credentials by credentials file path

### CODE REVIEW

File: column/api/backend/cache/local_mem.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+_key_queues = {}
+_locks = {}
+
+LOG = logging.getLogger(__name__)
+
+
+class LocalMemoryStore(store.Store):
+    """In memory store base class
+
+    This class inherits basic store class to implment local memory backend.
```
Comment (ericwb) [inline column/api/backend/cache/local_mem.py]: ...implement a....

File: column/api/backend/cache/__init__.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
+import local_mem
```
Comment (ericwb) [inline column/api/backend/cache/__init__.py]: always import full path.  should be:

from column.api.backend.cache import local_mem

File: column/api/backend/__init__.py [hunk 3]
```diff
@@ -0,0 +1,5 @@
+import cache
```
Comment (ericwb) [inline column/api/backend/__init__.py]: always import full path. should be:

from column.api.backend import cache

File: column/api/backend/cache/local_mem.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        return del_key
+
+    def _keys(self):
+        return self._store.keys()
+
+
+class RunMemoryStore(LocalMemoryStore):
+    """Column api service in-memory backend interface class
+
+    This class provides backend interface and additional logic for
```
Comment (ericwb) [inline column/api/backend/cache/local_mem.py]: provides a

File: column/api/backend/cache/local_mem.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    getting and saving run info
```
Comment (ericwb) [inline column/api/backend/cache/local_mem.py]: info.

File: column/api/backend/cache/local_mem.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+    """
+
+    def _get_progress(self, run):
+        return run['api_runner'].get_progress()
+
+    def _format_response(self, run):
+        # Get prgress only happens in playbook running
```
Comment (ericwb) [inline column/api/backend/cache/local_mem.py]: I don't understand this sentence.  It's a run-on sentence and has misspellings.

Get prgress only happens in playbook running after playbook is done, the api_runner attribute would be deleted

File: column/api/backend/cache/store.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+
+LOG = logging.getLogger(__name__)
+
+MAX_STORE_SIZE = 50
+
+
+class Store(object):
+    """Backend store base class
+
+    This is basic store implementation which is thread safe, local memory
```
Comment (ericwb) [inline column/api/backend/cache/store.py]: This is a.....

.......safe.  Local......

File: column/api/backend/cache/store.py [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+    store should inherit this class.
+
+    Attributes:
+        name (str): store name
+        lock (dict): dict of locks for multi-thread read and write operation
```
Comment (ericwb) [inline column/api/backend/cache/store.py]: multi-threaded..... operations

File: column/api/controller/credential_controller.py [hunk 9]
```diff
@@ -0,0 +1,35 @@
+import logging
+from six.moves import http_client
+
+from flask import request
+from flask_restful import abort
+from flask_restful import Resource
```
Comment (ericwb) [inline column/api/controller/credential_controller.py]: don't import classes only modules

File: column/api/controller/credential_controller.py [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+
+from column.api import manager
+
+LOG = logging.getLogger(__name__)
+
+
+class Credential(Resource):
+    """Column api credential resource class
+
+    Attributes:
```
Comment (ericwb) [inline column/api/controller/credential_controller.py]: This isn't proper docstring format.

File: column/api/controller/credential_controller.py [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+         manager (obj): manager interface for calling column library method to
+                        get/update the credential
+    """
+
+    def __init__(self):
+        self.manager = manager.get_manager('credential')
+
+    def get(self):
+        """Get credential by filepath"""
```
Comment (ericwb) [inline column/api/controller/credential_controller.py]: file path

File: column/api/controller/credential_controller.py [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+        cred_payload = request.get_json(silent=True)
+        cred = self.manager.get_credential(cred_payload)
+        return cred if cred else abort(http_client.NOT_FOUND,
```
Comment (ericwb) [inline column/api/controller/credential_controller.py]: This is an overly complex return statement.  Just use if-else block

File: column/api/controller/credential_controller.py [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+            message="Credential {} not found".format(cred['file_path']))
+
+    def put(self):
+        """Update credential by filepath"""
```
Comment (ericwb) [inline column/api/controller/credential_controller.py]: a credential

file path

File: column/api/controller/credential_controller.py (hunk 11 above)
Comment (ericwb) [inline column/api/controller/credential_controller.py]: a credential

File: column/api/controller/run_controller.py [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+        backend_store (obj): backend interface for retrieving run info
+    """
+
+    def __init__(self):
+        self.backend_store = backend.get_store()
+
+    def get(self, id):
+        """Get run by id"""
+        run = self.backend_store.get_run(id)
+        return run if run else abort(404,
```
Comment (ericwb) [inline column/api/controller/run_controller.py]: don't put complex if-else statements on a return like this

File: column/api/controller/run_controller.py [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+            message="Run {} doesn't exist".format(id))
+
+
+class RunList(Resource):
+    """Column api runlist resource class
+
+    Attributes:
```
Comment (ericwb) [inline column/api/controller/run_controller.py]: use proper docstring format

File: column/api/controller/run_controller.py [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+        self.manager = manager.get_manager('run')
+
+    def get(self):
+        """Get run list"""
+        LOG.info('Returning all ansible runs')
+        return self.backend_store.list_runs()
+
+    def post(self):
+        """Trigger a new run"""
+        # TODO(sihan) Add Run factory for validating the payload fields
```
Comment (ericwb) [inline column/api/controller/run_controller.py]: Will this be done in a future patch?  Also user id sihan doesn't exist in github

File: column/api/manager/credential_manager.py [hunk 17]
```diff
@@ -0,0 +1,21 @@
...
+
+    The Credential Manager layer is to support additional logic which is needed
+    to update and get credential.
+
+    """
+
+    def get_credential(self, cred):
+        return utils.get_vault_secret(cred['file_path'])
+
+    def update_credential(self, cred):
```
Comment (ericwb) [inline column/api/manager/credential_manager.py]: why is cred a dict? why not two parms?

File: column/api/manager/run_manager.py [hunk 18]
```diff
@@ -0,0 +1,79 @@
+import copy
+import logging
+import threading
+
+import column
+from column.api import backend
+from column.api.objects import State
```
Comment (ericwb) [inline column/api/manager/run_manager.py]: don't import classes, just modules

File: column/api/manager/run_manager.py [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+LOG = logging.getLogger(__name__)
+
+
+class RunManager(object):
+    """Column api manager class
+
+    The Manager layer is to support additional logic which is needed
+    to run ansible playbook.
+
+    Attributes:
```
Comment (ericwb) [inline column/api/manager/run_manager.py]: use docstring format

File: tests/api/controllers/test_credential.py [hunk 20]
```diff
@@ -0,0 +1,30 @@
...
+
+
+class TestCredential(controllers.APITest):
+
+    @patch('column.utils.get_vault_secret')
+    def test_get_credential(self, mock_get_vault_secret):
+        file_path = '/var/lib/credential/test_password'
+        mock_get_vault_secret.return_value = 'TestPassword'
+        response = self.app.get('/credentials', data=json.dumps(
+                                            dict(file_path=file_path,)),
```
Comment (ericwb) [inline tests/api/controllers/test_credential.py]: indentation way off

File: tests/api/controllers/test_credential.py [hunk 21]
```diff
@@ -0,0 +1,30 @@
...
+        self.assertTrue(mock_get_vault_secret.called)
+        mock_get_vault_secret.assert_called_with(file_path)
+        self.assertEqual(http_client.OK, response.status_code)
+
+    @patch('column.utils.update_vault_secret')
+    def test_update_credential(self, mock_update_vault_secret):
+        file_path = '/var/lib/credential/test_password'
+        value = 'TestPassword'
+        response = self.app.put('/credentials', data=json.dumps(
+                                    dict(file_path=file_path, value=value,)),
```
Comment (ericwb) [inline tests/api/controllers/test_credential.py]: indentiation off

File: tests/api/controllers/test_run.py [hunk 22]
```diff
@@ -0,0 +1,85 @@
...
+                break
+            time.sleep(self.wait_interval)
+            counter += 1
+
+        self.assertNotEqual('RUNNING', res_dict['state'])
+
+    def test_failed_run(self):
+        pb = 'tests/fixtures/playbooks/hello_world_with_fail.yml'
+        response = self.app.post('/runs', data=json.dumps(
+                                            dict(playbook_path=pb,
```
Comment (ericwb) [inline tests/api/controllers/test_run.py]: indentation is off

File: tests/api/controllers/test_run.py [hunk 23]
```diff
@@ -0,0 +1,85 @@
...
+        res_dict = json.loads(response.data)
+        self.assertEqual('ERROR', res_dict['state'])
+
+    def test_get_run_by_id(self):
+        response = self.app.get('/runs/1234')
+        self.assertEqual(http_client.NOT_FOUND, response.status_code)
+
+        pb = 'tests/fixtures/playbooks/hello_world.yml'
+        response = self.app.post('/runs', data=json.dumps(
+                                            dict(playbook_path=pb,
```
Comment (ericwb) [inline tests/api/controllers/test_run.py]: indentation is off

File: tests/api/controllers/test_run.py [hunk 24]
```diff
@@ -0,0 +1,85 @@
...
+        run_id = res_dict['id']
+        self.assertEqual(http_client.OK, response.status_code)
+        response = self.app.get('/runs/{}'.format(run_id))
+        self.assertEqual(http_client.OK, response.status_code)
+        self._wait_for_run_complete(run_id)
+
+    def test_get_run_list(self):
+        pb = 'tests/fixtures/playbooks/hello_world.yml'
+        response = self.app.post('/runs', data=json.dumps(
+                                            dict(playbook_path=pb,
```
Comment (ericwb) [inline tests/api/controllers/test_run.py]: indentation is off

File: tests/api/controllers/test_run.py [hunk 25]
```diff
@@ -0,0 +1,85 @@
...
+            if item['id'] == run_id:
+                found = True
+                break
+        self.assertEqual(True, found)
+        self._wait_for_run_complete(run_id)
+
+    def test_trigger_run(self):
+        pb = 'tests/fixtures/playbooks/hello_world.yml'
+        response = self.app.post('/runs', data=json.dumps(
+                                            dict(playbook_path=pb,
```
Comment (ericwb) [inline tests/api/controllers/test_run.py]: indentation is off

File: column/api/manager/run_manager.py [hunk 26]
```diff
@@ -0,0 +1,79 @@
...
+        failed_hosts = result['failed_hosts']
+        run = self.backend_store.get_run(run_id)
+        if unreachable_hosts:
+            error_msg = ("{} failed because the following nodes were "
+                         "unreachable: {}.\n".format(run['playbook_path'],
+                unreachable_hosts))
+            run['state'] = State.ERROR
+            run['message'] = error_msg
+            self.backend_store.update_run(run_id, run)
+            return
```
Comment (ericwb) [inline column/api/manager/run_manager.py]: This is a strange series of if-blocks.  Just do if, elif, else. No need to put empty return statements.

File: column/api/manager/run_manager.py (hunk 19 above)
Comment (ushergod) [inline column/api/manager/run_manager.py]: i am following http://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html

Do you have example what the format should be?

File: column/api/manager/credential_manager.py (hunk 17 above)
Comment (ushergod) [inline column/api/manager/credential_manager.py]: Cred is payload from client, it is dict. manager would take care of extracting params from payload.

File: tests/api/controllers/test_credential.py [hunk 27]
```diff
@@ -0,0 +1,30 @@
+import json
+from mock import patch
+
+from tests.api import controllers
+from six.moves import http_client
```
Comment (ericwb) [inline tests/api/controllers/test_credential.py]: alpha sort

File: tests/api/controllers/test_credential.py [hunk 28]
```diff
@@ -0,0 +1,30 @@
...
+
+
+class TestCredential(controllers.APITest):
+
+    @patch('column.utils.get_vault_secret')
```
Comment (ericwb) [inline tests/api/controllers/test_credential.py]: column.utils.vault_decrypt

File: tests/api/controllers/test_credential.py [hunk 29]
```diff
@@ -0,0 +1,30 @@
...
+        file_path = '/var/lib/credential/test_password'
+        mock_get_vault_secret.return_value = 'TestPassword'
+        response = self.app.get('/credentials', data=json.dumps(
+            dict(file_path=file_path,)),
+            content_type='application/json')
+        self.assertTrue(mock_get_vault_secret.called)
+        mock_get_vault_secret.assert_called_with(file_path)
+        self.assertEqual(http_client.OK, response.status_code)
+
+    @patch('column.utils.update_vault_secret')
```
Comment (ericwb) [inline tests/api/controllers/test_credential.py]: column.utils.vault_encrypt

File: tests/api/controllers/test_run.py [hunk 30]
```diff
@@ -0,0 +1,85 @@
+import json
+import time
+
+from tests.api import controllers
+from six.moves import http_client
```
Comment (ericwb) [inline tests/api/controllers/test_run.py]: alpha sort

File: column/api/run.py [hunk 31]
```diff
@@ -0,0 +1,29 @@
+import flask
+import flask_restful
+import logging
+
+
+from column import cfg
```
Comment (ericwb) [inline column/api/run.py]: alpha sort

File: column/api/controller/credential_controller.py [hunk 32]
```diff
@@ -0,0 +1,37 @@
+import logging
+from six.moves import http_client
```
Comment (ericwb) [inline column/api/controller/credential_controller.py]: six is 3rd party, should be grouped with flask

File: column/api/run.py (hunk 31 above)
Comment (ushergod) [inline column/api/run.py]: logging is basic package, i think we should not group logging into column

**Comment by ushergod** [comment]
We should focus on the https://github.com/vmware/column/pull/22 first, credential patch is based on that patch. That is why I create two pull request.

File: column/api/controller/run_controller.py [hunk 33]
```diff
@@ -0,0 +1,77 @@
+import logging
+from six.moves import http_client
```
Comment (ericwb) [inline column/api/controller/run_controller.py]: move six to 3rd party modules with flask

File: column/api/manager/credential_manager.py [hunk 34]
```diff
@@ -0,0 +1,21 @@
...
+from column import utils
+
+
+LOG = logging.getLogger(__name__)
+
+
+class CredentialManager(object):
+    """Column Credential Manager class
+
+    The Credential Manager layer is to support additional logic which is needed
```
Comment (ericwb) [inline column/api/manager/credential_manager.py]: This sentence doesn't say anything more than noting the names of obvious methods.

File: column/api/controller/credential_controller.py [hunk 35]
```diff
@@ -0,0 +1,33 @@
...
+from flask import request
+import flask_restful
+from six.moves import http_client
+
+from column.api import manager
+
+LOG = logging.getLogger(__name__)
+
+
+class Credential(flask_restful.Resource):
```
Comment (ericwb) [inline column/api/controller/credential_controller.py]: I don't think the naming of the class and methods make much sense anymore. We're not "getting" credentials.  We're getting an encrypted or decrypted form of a string.

File: column/api/common/utils.py [hunk 36]
```diff
@@ -0,0 +1,7 @@
+import cerberus
```
Comment (ericwb) [inline column/api/common/utils.py]: Why cerberus instead of jsonschema?

Comment (ushergod) [inline column/api/common/utils.py]: Both two are popular now. 
From my perspective, they are similar. Both ways are good.

Do you have any preference which way to go?

File: column/api/controller/credential_controller.py (hunk 35 above)
Comment (ushergod) [inline column/api/controller/credential_controller.py]: What about 'Secret'?

**Comment by ericwb** [comment]
Let's merge and fix issues afterwards.

**Pull request merged**